File: algorithms/visualstudio/algorithms.fsharp.tests/KaratsubaMultiplicationTests.fs
namespace algorithms.fSharp.divideAndConquer.tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open algorithms.fSharp.divideAndConquer.KaratsubaMultiplication
open System.Numerics

[<TestClass>]
type KaratsubaMultiplication() = 
    [<TestMethod>]
    member x.``KaratsubaMultiplication F# dividePoint 10 should return 1.`` () = 
        Assert.AreEqual(1, dividePoint 10I)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# dividePoint 5678 should return 2.`` () = 
        Assert.AreEqual(2, dividePoint 5678I)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# dividePoint 567898 should return 3.`` () = 
        Assert.AreEqual(3, dividePoint 567898I)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# dividePoint 1234567890123456789012345678901234567890 should return 20.`` () = 
        Assert.AreEqual(20, dividePoint 1234567890123456789012345678901234567890I)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# splitNumber 12 1 should return (1,2).`` () = 
        Assert.AreEqual((1I,2I), splitNumber 12I 1)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# splitNumber 1000 3 should return (1,0).`` () = 
        Assert.AreEqual((1I,0I), splitNumber 1000I 3)
    
    [<TestMethod>]
    member x.``KaratsubaMultiplication F# splitNumber 567898 4 should return (56,7898).`` () = 
        Assert.AreEqual((56I,7898I), splitNumber 567898I 4)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# splitNumber 1234567890123456789012345678901234567890I 20 should return (12345678901234567890,12345678901234567890).`` () = 
        Assert.AreEqual((12345687901234567890I,12345687901234567890I), splitNumber 1234568790123456789012345687901234567890I 20)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# 2 +*+? 5 should return 10.`` () = 
        Assert.AreEqual(10I, 2I +*+? 5I)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# 5678 +*+? 1234 should return 7006652.`` () = 
        Assert.AreEqual(7006652I, 5678I +*+? 1234I)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# 50 +*+? 10 should return 500.`` () = 
        Assert.AreEqual(500I, 50I +*+? 10I)

    [<TestMethod>]
    member x.``KaratsubaMultiplication F# 1548564532135665645648I +*+? 1548564554135665641248I should return 2398052144256972738312111343740159860488704I.`` () = 
        Assert.AreEqual(2398052144256972738312111343740159860488704I, 1548564532135665645648I +*+? 1548564554135665641248I)

File: algorithms/visualstudio/algorithms.csharp.tests/DivideAndConquer/KaratsubaMultiplicationTests.cs
using System.Numerics;
using algorithms.csharp.DivideAndConquer;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace algorithms.csharp.tests.DivideAndConquer
{
    [TestClass]
    public class KaratsubaMultiplicationTests
    {
        private static KaratsubaMultiplication _karatsuba;

        [ClassInitialize]
        public static void InitializeClass(TestContext ctx)
        {
            _karatsuba = new KaratsubaMultiplication();
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_DividePoint_10_Should_Return_1()
        {
            Assert.AreEqual(1, _karatsuba.DividePoint(MakeBigInteger(10)));
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_DividePoint_5678_Should_Return_2()
        {
            Assert.AreEqual(2, _karatsuba.DividePoint(MakeBigInteger(5678)));
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_DividePoint_567898_Should_Return_3()
        {
            Assert.AreEqual(3, _karatsuba.DividePoint(MakeBigInteger(567898)));
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_DividePoint_1234567890123456789012345678901234567890_Should_Return_20()
        {
            Assert.AreEqual(20, _karatsuba.DividePoint(MakeBigInteger("1234567890123456789012345678901234567890")));
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_SplitNumber_12_1_Should_Return_1_HighNumber()
        {
            Assert.AreEqual(MakeBigInteger(1), _karatsuba.SplitNumber(MakeBigInteger(12), 1).HighNumber);
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_SplitNumber_12_1_Should_Return_2_LowNumber()
        {
            Assert.AreEqual(MakeBigInteger(2), _karatsuba.SplitNumber(MakeBigInteger(12), 1).LowNumber);
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_SplitNumber_1000_3_Should_Return_1_HighNumber()
        {
            Assert.AreEqual(MakeBigInteger(1), _karatsuba.SplitNumber(MakeBigInteger(1000), 3).HighNumber);
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_SplitNumber_1000_3_Should_Return_0_LowNumber()
        {
            Assert.AreEqual(MakeBigInteger(0), _karatsuba.SplitNumber(MakeBigInteger(1000), 3).LowNumber);
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_SplitNumber_567898_4_Should_Return_56_HighNumber()
        {
            Assert.AreEqual(MakeBigInteger(56), _karatsuba.SplitNumber(MakeBigInteger(567898), 4).HighNumber);
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_SplitNumber_567898_4_Should_Return_7898_LowNumber()
        {
            Assert.AreEqual(MakeBigInteger(7898), _karatsuba.SplitNumber(MakeBigInteger(567898), 4).LowNumber);
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_SplitNumber_1234568790123456789012345687901234567890_20_Should_Return_12345687901234567890_HighNumber()
        {
            Assert.AreEqual(MakeBigInteger("12345687901234567890"), _karatsuba.SplitNumber(MakeBigInteger("1234568790123456789012345687901234567890"), 20).HighNumber);
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_SplitNumber_1234568790123456789012345687901234567890_20_Should_Return_12345687901234567890_LowNumber()
        {
            Assert.AreEqual(MakeBigInteger("12345687901234567890"), _karatsuba.SplitNumber(MakeBigInteger("1234568790123456789012345687901234567890"), 20).LowNumber);
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_Multiply_2_5_Should_Return_10()
        {
            Assert.AreEqual(MakeBigInteger(10), _karatsuba.Multiply(2,5));
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_Multiply_5678_1234_Should_Return_7006652()
        {
            Assert.AreEqual(MakeBigInteger(7006652), _karatsuba.Multiply(5678, 1234));
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_Multiply_50_10_Should_Return_500()
        {
            Assert.AreEqual(MakeBigInteger(500), _karatsuba.Multiply(50, 10));
        }

        [TestMethod]
        public void KaratsubaMultiplication_CSharp_Multiply_1548564532135665645648_1548564554135665641248_Should_Return_2398052144256972738312111343740159860488704()
        {
            Assert.AreEqual(MakeBigInteger("2398052144256972738312111343740159860488704"), _karatsuba.Multiply(MakeBigInteger("1548564532135665645648"), MakeBigInteger("1548564554135665641248")));
        }

        private BigInteger MakeBigInteger(int number)
        {
            return new BigInteger(number);
        }

        private BigInteger MakeBigInteger(string number)
        {
            return BigInteger.Parse(number);
        }
    }
}
